Introduce WKSRKEntity interface that wraps a RealityKit Entity object

achan00 opened the pull request:
#### bb0b0dc7700f6f8a68cbc4e5c9b79b1983916aba
<pre>
Introduce WKSRKEntity interface that wraps a RealityKit Entity object
<a href="https://bugs.webkit.org/show_bug.cgi?id=278337">https://bugs.webkit.org/show_bug.cgi?id=278337</a>
<a href="https://rdar.apple.com/134272371">rdar://134272371</a>

Reviewed by Mike Wyrzykowski.

This wrapper class allows us to manipulate the entity associated with
a model element using RealityKit APIs. This patch converts the logic
to get the entity&apos;s bounding box, get/set its transform and opacity,
and animate the entity to use RealityKit APIs in Swift.

* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
Add a new data member to hold a WKSRKEntity created with the REEntityRef
from the REModel.
m_animationPlaybackToken can be removed since we can now handle
animation with the RealityKit AnimationPlaybackController API.
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::updateTransform):
(WebKit::ModelProcessModelPlayerProxy::updateOpacity):
(WebKit::ModelProcessModelPlayerProxy::startAnimating):
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):
Create a WKSRKEntity with the CoreRE entity from REModel
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift: Added.
(WKSRKEntity.boundingBoxExtents):
(WKSRKEntity.boundingBoxCenter):
(WKSRKEntity.boundingBox):
(WKSRKEntity.transform):
(WKSRKEntity.opacity):
(WKSRKEntity.startAnimating):
* Source/WebKit/WebKitSwift/RealityKit/RealityKitBridging.h: Copied from Source/WebKit/WebKitSwift/WebKitSwift.h.
* Source/WebKit/WebKitSwift/WebKitSwift.h:

Canonical link: <a href="https://commits.webkit.org/282494@main">https://commits.webkit.org/282494@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/47caa113eebbf89cb4fbe5ab951fadd4b041031c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63287 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42643 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15884 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67308 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13895 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50331 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14175 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50979 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9594 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66356 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39590 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54801 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31662 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36273 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12767 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57812 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12478 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/69004 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7234 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12083 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58291 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7265 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54871 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58517 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14024 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6024 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38464 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39543 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40655 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/39286 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eddydas commented on Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift:
```diff
@@ -0,0 +1,101 @@
+        get {
+            guard let transformComponent = entity.components[Transform.self] else {
+                Logger.realityKitEntity.error("No transform component available from entity")
+                return WKEntityTransform(scale: simd_float3.one, rotation: simd_quatf(ix: 0, iy: 0, iz: 0, r: 1), translation: simd_float3.zero)
```
Nit: Is there a const for no-rotation quat?

achan00 replied:
I've seen other frameworks define their own. If we have more instances of this, we can too.

eddydas commented on Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift:
```diff
@@ -0,0 +1,101 @@
+        }
+
+        set {
+            if newValue >= 1.0 {
```
Should we simply clamp the opacity to [0,1] ?
Also, missing handling for <0 ?

achan00 replied:
The clamping currently happened in the caller. I can move it here. I just converted the CoreRE logic from `ModelProcessModelPlayerProxy` to here.

eddydas commented on Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift:
```diff
@@ -0,0 +1,101 @@
+        entity = Entity.__fromCore(__EntityRef.__fromCore(coreEntity))
+    }
+
+    @objc(boundingBoxExtents) public var boundingBoxExtents: simd_float3 {
```
Do we want to return a nullable here? (if it's something that can be supported when bridging into ObjC?)
Same for boundingBoxCenter

achan00 replied:
Why do you think this needs to be nullable?

eddydas replied:
Just because the underlying boundingBox object is nullable.

achan00 replied:
I see - yeah, I think if I want to this to be able to return null, I'd probably have to return a `simd_float3*`, and if I do that, then the caller would have to handle the null result somehow. Maybe simpler to keep it the way it is.

achan00 committed:
Introduce WKSRKEntity interface that wraps a RealityKit Entity object
https://bugs.webkit.org/show_bug.cgi?id=278337
rdar://134272371

Reviewed by Mike Wyrzykowski.

This wrapper class allows us to manipulate the entity associated with
a model element using RealityKit APIs. This patch converts the logic
to get the entity's bounding box, get/set its transform and opacity,
and animate the entity to use RealityKit APIs in Swift.

* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
Add a new data member to hold a WKSRKEntity created with the REEntityRef
from the REModel.
m_animationPlaybackToken can be removed since we can now handle
animation with the RealityKit AnimationPlaybackController API.
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::updateTransform):
(WebKit::ModelProcessModelPlayerProxy::updateOpacity):
(WebKit::ModelProcessModelPlayerProxy::startAnimating):
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):
Create a WKSRKEntity with the CoreRE entity from REModel
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift: Added.
(WKSRKEntity.boundingBoxExtents):
(WKSRKEntity.boundingBoxCenter):
(WKSRKEntity.boundingBox):
(WKSRKEntity.transform):
(WKSRKEntity.opacity):
(WKSRKEntity.startAnimating):
* Source/WebKit/WebKitSwift/RealityKit/RealityKitBridging.h: Copied from Source/WebKit/WebKitSwift/WebKitSwift.h.
* Source/WebKit/WebKitSwift/WebKitSwift.h:

Canonical link: https://commits.webkit.org/282494@main
